pkg/crypto/tinkcrypto/primitive/composite/ecdh/ecdh_xchacha_aead_private_key_manager.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/hyperledger/aries-framework-go/pkg/internal/cryptoutil"
+)
+
+const (
+	ecdhXChachaPrivateKeyVersion = 0
+	ecdhXChachaPrivateKeyTypeURL = "type.hyperledger.org/hyperledger.aries.crypto.tink.EcdhXChachaAeadPrivateKey"
+)
+
+// common errors.
+var (
+	errInvalidECDHXChachaPrivateKey       = errors.New("ecdh_xchacha_private_key_manager: invalid key")
+	errInvalidECDHXChachaPrivateKeyFormat = errors.New("ecdh_xchacha_private_key_manager: invalid key format")
+)
+
+// ecdhXChachaPrivateKeyManager is an implementation of PrivateKeyManager interface.
+// It generates new ECDHPrivateKey (XChacha) keys and produces new instances of ECDHAEADCompositeDecrypt subtle.
+type ecdhXChachaPrivateKeyManager struct{}
+
+// Assert that ecdhXChachaPrivateKeyManager implements the PrivateKeyManager interface.
+var _ registry.PrivateKeyManager = (*ecdhXChachaPrivateKeyManager)(nil)
+
+// newECDHXChachaPrivateKeyManager creates a new ecdhXChachaPrivateKeyManager.
+func newECDHXChachaPrivateKeyManager() *ecdhXChachaPrivateKeyManager {
+	return new(ecdhXChachaPrivateKeyManager)
+}
+
+// Primitive creates an ECDHESPrivateKey subtle for the given serialized ECDHESPrivateKey proto.
+func (km *ecdhXChachaPrivateKeyManager) Primitive(serializedKey []byte) (interface{}, error) {
+	if len(serializedKey) == 0 {
+		return nil, errInvalidECDHXChachaPrivateKey
+	}
+
+	key := new(ecdhpb.EcdhAeadPrivateKey)
+
+	err := proto.Unmarshal(serializedKey, key)
+	if err != nil {
+		return nil, errInvalidECDHXChachaPrivateKey
+	}
+
+	err = km.validateKey(key)
+	if err != nil {
+		return nil, errInvalidECDHXChachaPrivateKey
+	}
+
+	rEnc, err := composite.NewRegisterCompositeAEADEncHelper(key.PublicKey.Params.EncParams.AeadEnc)
+	if err != nil {
+		return nil, fmt.Errorf("ecdh_xchacha_private_key_manager: NewRegisterCompositeAEADEncHelper "+
+			"failed: %w", err)
+	}
+
+	return subtle.NewECDHAEADCompositeDecrypt(rEnc, key.PublicKey.Params.EncParams.CEK), nil
+}
+
+// NewKey creates a new key according to the specification of ECDHESPrivateKey format.
+func (km *ecdhXChachaPrivateKeyManager) NewKey(serializedKeyFormat []byte) (proto.Message, error) {
+	if len(serializedKeyFormat) == 0 {
+		return nil, errInvalidECDHXChachaPrivateKeyFormat
+	}
+
+	keyFormat := new(ecdhpb.EcdhAeadKeyFormat)
+
+	err := proto.Unmarshal(serializedKeyFormat, keyFormat)
+	if err != nil {
+		return nil, errInvalidECDHXChachaPrivateKeyFormat
+	}
+
+	err = validateKeyXChachaFormat(keyFormat.Params)
+	if err != nil {
+		return nil, errInvalidECDHXChachaPrivateKeyFormat
+	}
+
+	pub, pvt, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("ecdh_xchacha_private_key_manager: GenerateECDHKeyPair failed: %w", err)
+	}
+
+	x25519Pub, err := cryptoutil.PublicEd25519toCurve25519(pub)
+	if err != nil {
+		return nil, fmt.Errorf("ecdh_xchacha_private_key_manager: Convert to X25519 pub key failed: %w", err)
+	}
+
+	x25519Pvt, err := cryptoutil.SecretEd25519toCurve25519(pvt)
```

Comment:
Is there no function to just generate X25519 directly?

Reply:
I haven't found one, I guess since it's a random []byte sequence, the closest I've seen is [HChacha()](https://github.com/golang/crypto/blob/master/chacha20/chacha_generic.go#L344), but it creates a key from another key and a nonce which is used during Chacha primitive execution.